### PR TITLE
Adding DataLoader2 Adapter API and shuffle adapter

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #485
* __->__ #484

Differential Revision: [D36866291](https://our.internmc.facebook.com/intern/diff/D36866291)

### CODE REVIEW

File: torchdata/dataloader2/adapter.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    "shuffle",
+]
+
+assert __all__ == sorted(__all__)
+
+Adapter = Callable[[IterDataPipe], IterDataPipe]
+
+
+def shuffle(enable=None) -> Adapter:
+    return functools.partial(_shuffle, enable)
```
Comment (VitalyFedyunin) [inline torchdata/dataloader2/adapter.py:25]: TODO: Inline docs

**Review by NivekT — APPROVED** [review]

File: torchdata/dataloader2/adapter.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from torchdata.datapipes.iter import IterDataPipe
+
+__all__ = [
+    "Adapter",
+    "shuffle",
+]
+
+assert __all__ == sorted(__all__)
+
+Adapter = Callable[[IterDataPipe], IterDataPipe]
```
Comment (ejguan) [inline torchdata/dataloader2/adapter.py:21]: Do you think it's better to provide a base class of callable `Adapter` ?
```py
class Adapter:
    @abstractmethod
    def __call__(self, datapipe: DataPipe) -> DataPipe:
        pass
```
- It's easier to check `isinstance(adapter, Adapter)`
- The type checking might be easier

Comment (VitalyFedyunin) [reply torchdata/dataloader2/adapter.py:21]: Surely it is. I just blanked out about this option. Thanks.

File: torchdata/dataloader2/dataloader2.py [hunk 3]
```diff
@@ -45,22 +47,27 @@ class DataLoader2(Generic[T_co]):
...
         self._datapipe_iter: Optional[Iterator[T_co]] = None
         self._reset_iter: bool = True
         # TODO(VitalyFedyunin): Some ReadingServices might want to validate adapters, we can add this feature
-        self.datapipe_adapter_fn = datapipe_adapter_fn
+        if datapipe_adapter_fn is None:
+            self.datapipe_adapter_fns = None
+        elif isinstance(datapipe_adapter_fn, Iterable):
+            self.datapipe_adapter_fns = datapipe_adapter_fn
+        else:
+            self.datapipe_adapter_fns = [datapipe_adapter_fn]
```
Comment (ejguan) [inline torchdata/dataloader2/dataloader2.py:63]: Do we want to add a check here to validate all `fns` is an `Adapter`.

File: torchdata/dataloader2/adapter.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+class Adapter:
+    @abstractmethod
+    def __call__(self, datapipe: IterDataPipe) -> IterDataPipe:
+        pass
+
+
+class Shuffle(Adapter):
+    r"""
+    Shuffle DataPipes adapter allows control over all existing Shuffler (`shuffle`) DataPipes in the graph.
```
Comment (NivekT) [inline torchdata/dataloader2/adapter.py:29]:
```suggestion
    Shuffle DataPipes adapter allows control over all existing Shuffler (``shuffle``) DataPipes in the graph.
```
Nit: need double backticks for docstrings (that's how sphinx is)

**Review by NivekT — APPROVED** [review]
Nit comments and questions

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36866291).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36866291).

**Review by Miiira — APPROVED** [review]
LGTM!

**Pull request closed without merging**